application-ideas-api/src/main/java/com/xwiki/ideas/internal/DefaultIdeasManager.java

```diff
@@ -0,0 +1,189 @@
+                .append(URLEncoder.encode(documentReference.getWikiReference().getName(), XWiki.DEFAULT_ENCODING));
+            for (SpaceReference spaceReference : documentReference.getSpaceReferences()) {
+                stringBuilder.append("/spaces/");
+                stringBuilder.append(URLEncoder.encode(spaceReference.getName(), XWiki.DEFAULT_ENCODING));
+            }
+            stringBuilder.append("/pages/");
+            stringBuilder.append(URLEncoder.encode(documentReference.getName(), XWiki.DEFAULT_ENCODING));
+            stringBuilder.append("/idea");
+            return stringBuilder.toString();
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(String.format("Failed to retrieve the REST URL of the document: [%s]",
+                documentReference), e);
+        }
+    }
+
+    @Override public String vote(DocumentReference documentReference, String voteType) throws IdeasException
+    {
+        XWikiContext xcontext = contextProvider.get();
+        XWiki xWiki = xcontext.getWiki();
+        Map<String, Object> result = new HashMap<>();
+        try {
+            XWikiDocument mydoc = xWiki.getDocument(documentReference, xcontext);
+            BaseObject ideasObj = mydoc.getXObject(IDEA_CLASS_REFERENCE);
+            XWikiUser user = xWiki.checkAuth(xcontext);
+
+            if (!mydoc.isNew() && null != ideasObj) {
+                Map<String, Boolean> doRemoveMap = new HashMap<>();
+                doRemoveMap.put(VOTERS_FOR_KEY, false);
+                doRemoveMap.put(VOTERS_AGAINST_KEY, false);
+                // Action : Add a supporter (vote)
+                if (voteType.equals("addVote")) {
+                    addVote(VOTERS_FOR_KEY, VOTERS_AGAINST_KEY, ideasObj, user, result, xcontext, doRemoveMap);
+                } else if (voteType.equals("addVoteOpp")) {
+                    addVote(VOTERS_AGAINST_KEY, VOTERS_FOR_KEY, ideasObj, user, result, xcontext, doRemoveMap);
+                }
+                // Action : Remove a supporter (vote)
+                if (voteType.equals("removeVote") || doRemoveMap.get(VOTERS_FOR_KEY)) {
+                    // Remove user from user list
+                    decrementVote(VOTERS_FOR_KEY, ideasObj, user, xcontext, result);
+                } else if (doRemoveMap.get(VOTERS_AGAINST_KEY)) {
+                    // Remove user from user list
+                    decrementVote(VOTERS_AGAINST_KEY, ideasObj, user, xcontext, result);
+                }
+                // Save document
+                if (!result.isEmpty()) {
+                    xWiki.saveDocument(mydoc, "New Vote", xcontext);
+                }
+            } else {
+                throw new IdeasException(String.format("Failed to vote for [%s] on behalf of [%s].",
+                    documentReference, user.getUserReference()));
+            }
+            JSONTool jsonTool = new JSONTool();
+            return jsonTool.serialize(result);
```

Comment:
You need to return some typed object. Moreover, if your check the package of ``JSONTool`` you'll see that it's a Velocity tool, but the Ideas Manager has nothing to do with Velocity.